notes/connection-lifecycle.md

```diff
@@ -0,0 +1,77 @@
+
+Set the connection's `timeout` parameter to the "connect timeout" value.
+
+Connect to the first origin by calling the `HTTPConnection.connect()` method.
+If an error occurs here you can check whether the error occurred during the
+connection to the proxy if `HTTPConnection._connecting_to_proxy` is true.
+If the value is false then the error didn't occur while connecting to a proxy.
+
+(Should we be setting `timeout` to some write timeout at this point? `None`?)
+
+After connecting to the origin, the connection can be checked to see if `is_verified` is set to true. If not the pool emits a warning.
+The warning only matters for when verification is disabled, because otherwise an error is raised on unverified TLS handshake.
+
+Then the HTTP request can be sent with `HTTPConnection.request()`. If a `BrokenPipeError` (should we include `SSLEOFError`?)
+is raised while sending the request body it can be swallowed as a response
+can still be received from the origin even when the request isn't completely
+sent.
+
+At this point the `timeout` is set to the `read_timeout` (currently done through `conn.sock.settimeout()`, can we change this to `HTTPConnection.timeout`?)
+
+Then response headers (and other info) are read from the connection via `HTTPConnection.getresponse()` and returned as a `urllib3.HTTPResponse`. The `HTTPResponse` instance carries a reference to the `HTTPConnection` instance. (Within `HTTPConnectionPool` the `HTTPResponse._connection` is set to `response_conn`, is that needed or can `HTTPConnection.getresponse()` handle this?)
+
+If pooling is in use set `_connection` and `_pool` on the `HTTPResponse` instance. If retries are in use set `retries` on the `HTTPResponse` instance.
+
+If any error is received from connecting to the origin, sending the request, or receiving the response, the pool will call `HTTPConnection.close()` and discard the connection.
+
+### HTTP/HTTPS/proxies
+
+- No proxy, HTTP origin -> `HTTPConnection`
+- No proxy, HTTPS origin -> `HTTPSConnection`
+- HTTP proxy, HTTP origin -> `HTTPConnection` in forwarding mode
+- HTTP proxy, HTTPS origin -> `HTTPSConnection` in tunnel mode
+- HTTPS proxy, HTTP origin -> `HTTPSConnection` in forwarding mode
+- HTTPS proxy, HTTPS origin -> `HTTPSConnection` in tunnel mode
+- HTTPS proxy, HTTPS origin, `use_forwarding_for_https=True` -> `HTTPSConnection` in forwarding mode
+
+### Downsides
+
+- Lots of internal property usage by connection pools/response.
+  - Checking whether a connection is established
+  - Accessing `.sock` property
+- Proxies seem super clunky to setup even though we're giving all the information to the connection pool/connection?
+- Timeouts are getting set throughout an HTTPConnection's lifecycle by the connection pool
+- Does `HTTPConnection.set_tunnel()` give us anything? Remove? No-op?
+
+## How could things be better?
```

Comment:
This is most probably irrelevant, but I wanted to mention that having a pool-of-pools (ie. having both PoolManager and HTTPConnectionPool) is weird.

Reply:
Totally weird, we need to create a better recipe for a single connection pool of mixed-config connections.